contracts/Removal.sol

```diff
@@ -316,6 +335,85 @@ contract Removal is
     });
   }
 
+  /**
+   * @notice Transfers the provided `amounts` of the specified removal `ids` directly to the Certificate contract to
+   * mint a legacy certificate.
+   *
+   * @dev The Certificate contract implements `onERC1155BatchReceived`, which is invoked upon receipt of any removals.
+   * This function circumvents the market contract's lifecycle. This function first transfers the removals from the
+   * supplier accounts to a single account that has the consignor role-- this is necessary because
+   * `_safeBatchTransferFrom` only accepts one `from` address and `Certificate.onERC1155BatchReceived` will mint a *new*
+   * certificate every time an additional batch is received.
+   *
+   * ##### Requirements:
+   * - Can only be used when the caller has the `CONSIGNOR_ROLE`
+   * - Can only be used when this contract is not paused
+   * - IDs must already have been minted via `mintBatch`.
+   * - Enforces the rules of `Removal._beforeTokenTransfer`.
+   *
+   * @param owners The owners of the specified removal IDs to transfer from.
+   * @param ids 2D array of the removal IDs to add to transfer to the Certificate contract. Each row in the array
+   * represents a list of one owners removal IDs.
+   * @param amounts 2D array of the balances of each token ID to transfer to the Certificate contract. Each row in the
+   * array represents a list of one owner's removal amounts, each of which corresponds to the same row and column
+   * defined in `ids`.
+   * @param certificateRecipient The recipient of the certificate to be minted.
+   * @param certificateAmount TThe total amount of the certificate.
+   */
+  function migrate(
+    address[] calldata owners,
+    uint256[][] calldata ids,
+    uint256[][] calldata amounts,
+    address certificateRecipient,
+    uint256 certificateAmount
+  ) external whenNotPaused onlyRole(CONSIGNOR_ROLE) {
+    uint256 numberOfRemovals = 0;
+    // Skip overflow check as for loop is indexed starting at zero.
+    unchecked {
+      for (uint256 i = 0; i < owners.length; ++i) {
+        // Batch transfers removals from the supplier to the consignor
+        _safeBatchTransferFrom({
+          from: owners[i],
+          to: _msgSender(),
+          ids: ids[i],
+          amounts: amounts[i],
+          data: ""
+        });
+        numberOfRemovals += ids[i].length;
+      }
+    }
+    uint256[] memory flattenedAmounts = new uint256[](numberOfRemovals);
+    uint256[] memory flattenedIds = new uint256[](numberOfRemovals);
+    uint256 removalsTotal = 0;
+    // Skip overflow check as for loop is indexed starting at zero.
+    unchecked {
+      for (uint256 i = 0; i < flattenedAmounts.length; ++i) {
```

Comment:
I think `i` will get too big here to be using as an index to `amounts` right?  `i` spans the full number of removals (as it is iterating to `flattenedAmounts.length`, but `amounts` only has a few subarrays

Reply:
Good catch. Tests were only accounting for a [3][1] array so this wasn't an issue there. Will adjust the tests to use a better test data set that will catch that kind of thing (e.g. [3][2])